src/js/game.js

```diff
@@ -162,145 +212,158 @@ function Game() {
 		}
-		
+
 		this.update = function() {
 			if (this.direction === 'left') {
 				x -= this.speed;
-				
+
 				if (x-r < 0) {
 					this.direction = false;
 					x = r;
 				}
 			}
-			
+
 			if (this.direction === 'right') {
 				x += this.speed;
-				
+
 				if (x+r > gameW) {
 					this.direction = false;
 					x = gameW-r;
 				}
 			}
-			
+
 			draw();
 		}
-		
+
 		draw();
 	}
-	
-	function Ball() {
-		var r = 1.5e-2*gameW,
-			x = gameW/2,
-			y = gameH - gameH/100*5,
+
+	function Ball(initX, initY, r, speed) {
+		var x = initX,
+			y = initY,
 			teta = Math.random()*4*pi/6 + pi/6;
 
-		this.speed = 1e-2*gameH;
-			
+		this.x = x;
+		this.y = y;
+		this.teta = teta;
+		this.speed = speed;
+
 		function draw() {
 			c.fillStyle = 'lightgrey';
 			c.beginPath();
 			c.arc(x, y, r, 0, pi*2);
 			c.closePath();
 			c.fill();
-			
-			/* DEBUG */
+
+			/* DEBUG
 			c.fastLine(x, 0, x, gameH, 'lightgrey', 2);
 			c.fastLine(0, y, gameW, y, 'lightgrey', 2);
 			c.fillText('## DEBUG MODE ##', 50, gameH - 200);
 			c.fillText('θ = ' + (teta/pi*180).toFixed(3) + '°', 50, gameH - 180);
 			c.fillText('m = ' + Math.tan(teta).toFixed(3), 50, gameH - 170);
-			c.fastLine(x, y, x + 1000*Math.cos(-teta), y + 1000*Math.sin(-teta), 'pink', 2);
+			c.fastLine(x, y, x + 1000*Math.cos(-teta), y + 1000*Math.sin(-teta), 'pink', 2);*/
 		}
-			
+
 		this.update = function() {
 			// Walls
-			if (x+r >= gameW || x-r <= 0) teta = teta <= pi ? pi - teta : 3*pi - teta;	// Right and left walls
-			if (y-r <= 0) teta = 2*pi - teta;											// Top wall
-			if (pad.collision(x, y, r, teta)) teta = pad.newTeta;						// Pad
-			
+			if (x+r >= gameW || x-r <= 0) teta = Math.reflect(Infinity, teta);		// Right and left walls
+			if (y-r <= 0) teta = Math.reflect(0, teta);								// Top wall
+			if (pad.collision(x, y, r, teta)) teta = pad.newTeta;					// Pad
+			/* DEBUG	if (y+r >= gameH) teta = Math.reflect(0, teta);				// Make pad unnecessary */
+
 			// Bricks
-			for (var i=0; i < bricks.length; i++) {
-				for (var j=0, b; b = bricks[i][j]; j++) if (b.collision(x, y, r, teta)) {
-					teta = b.newTeta;
-					if (b.dead) bricks[i, j] = false;
+			if (y <= Ytreshold) // start checking for collisions with bricks only when the ball is close to the lowest one
+			for (var i=bricks[0].length-1, found=false; i >= 0; i--) { // loop optimization: start from lower ones (bottom)
+				for (var j=bricks.length-1; j >= 0; j--) {
+					if (bricks[i][j] && bricks[i][j].collision(x, y, r, teta)) {
+						teta = Math.reflect(bricks[i][j].tan, teta);
+						found = true;
+						break; // loop optimization: don't bother checking the other bricks after collision
+					}
 				}
+				if (found) break;
 			}
-			
+
 			// Canvases have inverted Y axis
-			y -= this.speed*Math.sin(teta);
-			x += this.speed*Math.cos(teta);
-			
+			this.y = y = (y - this.speed*Math.sin(teta)).limitTo(r,gameH+r);
+			this.x = x = (x + this.speed*Math.cos(teta)).limitTo(r,gameW-r);
+			this.teta = teta;
+
```

Comment:
This fix prevents the ball from going outside the canvas (causing it to bounce in loop through the border).